builder/docker/communicator.go

```diff
@@ -60,40 +60,83 @@ func (c *Communicator) Start(remote *packer.RemoteCmd) error {
 	return nil
 }
 
+// Upload uploads a file to the docker container
 func (c *Communicator) Upload(dst string, src io.Reader, fi *os.FileInfo) error {
+	if fi == nil {
+		return c.uploadReader(dst, src)
+	}
+	return c.uploadFile(dst, src, fi)
+}
+
+// uploadReader writes an io.Reader to a temporary file before uploading
+func (c *Communicator) uploadReader(dst string, src io.Reader) error {
 	// Create a temporary file to store the upload
 	tempfile, err := ioutil.TempFile(c.HostDir, "upload")
-	if err != nil {
-		return err
-	}
 	defer os.Remove(tempfile.Name())
+	defer tempfile.Close()
```

Comment:
Will moving this to the beginning cause issues if the error occured before the file was ever opened?